File: packages/kits/AriaMcpKit/rust/src/recall_discrimination.rs
//! Recall discrimination / confidence metric.
//!
//! A pure, scale-independent helper that classifies how well a ranked recall
//! result distinguishes its top hit from the rest of the list.
//!
//! This is a CONFIDENCE HEURISTIC, not a ranking signal. It is computed from
//! the scores the recall tools ALREADY have and is appended to the tool result
//! so the calling AI can interpret low-discrimination results correctly (treat
//! as effectively unranked, prefer lexical modes) rather than acting on noise.
//!
//! ## Parity contract
//!
//! Swift and Rust must produce identical [`DiscriminationLevel`] for the same
//! score vector. The named threshold constants here are mirrored verbatim in
//! `RecallDiscrimination.swift`. Any threshold change must be applied to both
//! ports simultaneously.
//!
//! ## Algorithm (scale-independent)
//!
//! Given ordered scores s0 ≥ s1 ≥ … ≥ s_{n-1}:
//!   - n < 2                                    → `Single`
//!   - topGap = (s0 - s1) / max(|s0|, EPS)
//!   - spread  = (s0 - sLast) / max(|s0|, EPS)
//!   - `Low`    if topGap < LOW_MARGIN && spread < LOW_SPREAD
//!   - `High`   if topGap >= HIGH_MARGIN
//!   - else     `Medium`
//!
//! Thresholds are heuristic defaults — tune with real estate data; do NOT rank
//! with these values.

// ---------------------------------------------------------------------------
// Threshold constants (mirrored in RecallDiscrimination.swift)
// ---------------------------------------------------------------------------

/// Minimum relative gap (rank-1 vs rank-2) for a "not low" classification.
const LOW_MARGIN: f64 = 0.05;
/// Minimum relative spread (rank-1 vs last) for a "not low" classification.
const LOW_SPREAD: f64 = 0.15;
/// Minimum relative gap (rank-1 vs rank-2) for a "high" classification.
const HIGH_MARGIN: f64 = 0.25;
/// Epsilon prevents division by zero on all-zero score vectors.
const EPS: f64 = 1e-9;

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/// How well a ranked recall result separates its top hit from the rest.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DiscriminationLevel {
    /// Fewer than two results — nothing to compare.
    Single,
    /// Top result is clearly separated from the second (topGap >= HIGH_MARGIN).
    High,
    /// Partial separation — some evidence of a best hit.
    Medium,
    /// Top results are within epsilon — the list is effectively unranked.
    Low,
}

// ---------------------------------------------------------------------------
// Classification
// ---------------------------------------------------------------------------

/// Classify the discrimination level of an ordered score list.
///
/// `scores` must already be in descending order (highest first). This function
/// does NOT sort the input.
pub fn classify(scores: &[f64]) -> DiscriminationLevel {
    if scores.len() < 2 {
        return DiscriminationLevel::Single;
    }

    let s0 = scores[0];
    let s1 = scores[1];
    let s_last = scores[scores.len() - 1];

    // Denominator: magnitude of the top score, floored at EPS so a
    // zero-top-score list still produces well-defined ratios (all ratios
    // become 0 / EPS ≈ 0 → Low).
    let denom = s0.abs().max(EPS);
    let top_gap = (s0 - s1) / denom;
    let spread = (s0 - s_last) / denom;

    if top_gap >= HIGH_MARGIN {
        DiscriminationLevel::High
    } else if top_gap < LOW_MARGIN && spread < LOW_SPREAD {
        DiscriminationLevel::Low
    } else {
        DiscriminationLevel::Medium
    }
}

// ---------------------------------------------------------------------------
// Result line
// ---------------------------------------------------------------------------

/// The AI-facing discrimination line to append to a ranked recall result.
///
/// Wording is factual and action-oriented so the calling AI knows what to do,
/// not just what the level is.
pub fn result_line(level: DiscriminationLevel) -> &'static str {
    match level {
        DiscriminationLevel::Single => "discrimination: n/a — single/zero results.",
        DiscriminationLevel::High => "discrimination: high — clear top result.",
        DiscriminationLevel::Medium => "discrimination: medium — partial separation.",
        // For small corpora the semantic/associative modes (conceptual shaped-recall,
        // partial_cue) produce near-flat scores until the embedding encoder (v1.1)
        // lands — this is expected, not an error. Direct the AI toward lexical/precise
        // modes when it needs ranking it can trust on small estates.
        DiscriminationLevel::Low => {
            "discrimination: low — top results are within epsilon; treat as effectively unranked. \
             Prefer moot_recall_precise / moot_memory_search (ordering: byRelevanceDesc) for \
             precision, or widen the query."
        }
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn single_or_zero_results_return_single() {
        assert_eq!(classify(&[]), DiscriminationLevel::Single);
        assert_eq!(classify(&[1.0]), DiscriminationLevel::Single);
    }

    #[test]
    fn clearly_separated_scores_return_high() {
        // topGap = (1.0 - 0.5) / 1.0 = 0.5 >= HIGH_MARGIN (0.25)
        assert_eq!(classify(&[1.0, 0.5, 0.3]), DiscriminationLevel::High);
    }

    #[test]
    fn near_flat_scores_return_low() {
        // topGap = (1.0 - 0.98) / 1.0 = 0.02 < LOW_MARGIN (0.05)
        // spread = (1.0 - 0.95) / 1.0 = 0.05 < LOW_SPREAD (0.15)
        assert_eq!(classify(&[1.0, 0.98, 0.97, 0.95]), DiscriminationLevel::Low);
    }

    #[test]
    fn medium_gap_returns_medium() {
        // topGap = (1.0 - 0.88) / 1.0 = 0.12  (>= LOW_MARGIN but < HIGH_MARGIN)
        assert_eq!(classify(&[1.0, 0.88, 0.50]), DiscriminationLevel::Medium);
    }

    #[test]
    fn all_zero_scores_return_low() {
        // denom = EPS, topGap ≈ 0 / EPS ≈ 0 < LOW_MARGIN, spread ≈ 0 → Low
        assert_eq!(classify(&[0.0, 0.0, 0.0]), DiscriminationLevel::Low);
    }

    #[test]
    fn two_items_high_gap_returns_high() {
        // topGap = (0.9 - 0.1) / 0.9 ≈ 0.89 >= HIGH_MARGIN
        assert_eq!(classify(&[0.9, 0.1]), DiscriminationLevel::High);
    }

    /// Parity: the same score vectors must produce the same level as Swift.
    /// This is a compile-time-verified cross-port contract: if you change a
    /// threshold here, update RecallDiscrimination.swift to match.
    #[test]
    fn parity_vectors_match_swift() {
        // Vector A: high separation → High
        assert_eq!(classify(&[1.0, 0.6]), DiscriminationLevel::High);
        // Vector B: flat spread, tiny gap → Low
        assert_eq!(classify(&[1.0, 0.99, 0.98, 0.97]), DiscriminationLevel::Low);
        // Vector C: medium separation → Medium
        assert_eq!(classify(&[1.0, 0.9, 0.4]), DiscriminationLevel::Medium);
    }
}

File: packages/kits/AriaMcpKit/Sources/AriaMCP/RecallDiscrimination.swift
// RecallDiscrimination.swift
//
// A pure, scale-independent helper that classifies how well a ranked recall
// result distinguishes its top hit from the rest of the list.
//
// This is a CONFIDENCE HEURISTIC, not a ranking signal. It is computed from
// the scores the recall tools ALREADY have and is appended to the tool result
// so the calling AI can interpret low-discrimination results correctly (treat
// as effectively unranked, prefer lexical modes) rather than acting on noise.
//
// The metric is intentionally lightweight: one relative gap ratio from rank-1
// to rank-2, and a spread ratio from rank-1 to the last result. Both are
// relative (divided by max(|s0|, EPS)) so the classification is
// score-scale-agnostic — it applies whether scores are in [0, 1] or [0, 1000].
//
// ## Dense-lane dark cap
//
// When the semantic vector lane (Lane D) is dark — i.e. the recall result
// carried `denseLaneStatus != nil` — the ranking is lexical/BM25 only.
// A pure-lexical ranking CAN produce a high score-gap (e.g. one memory
// contains the exact query token, others do not), but reporting "high — clear
// top result" while the semantic lane was unavailable violates the
// discrimination signal's own contract: it is supposed to tell the caller
// whether the ranking is TRUSTWORTHY. BM25-only rankings on small corpora are
// not semantically trustworthy. When the dense lane is dark the result is
// capped at .medium and a caveat is appended to the result line so the calling
// AI knows the semantic lane did not contribute.
//
// The cap is applied by `resultLine(for:denseLaneDark:)`. The plain
// `resultLine(for:)` overload (no dark-lane parameter) is preserved for
// callers that do not have dense-lane context (e.g. `recall_shaped`).
//
// Parity contract: Swift and Rust must produce identical DiscriminationLevel
// for the same score vector. The thresholds are named constants here and
// mirrored verbatim in `recall_discrimination.rs`. The dense-lane-dark cap
// and caveat wording are also mirrored; any change must be applied to both
// ports simultaneously.

import Foundation

/// How well a ranked recall result separates its top hit from the rest.
public enum DiscriminationLevel: Sendable, Equatable {
    /// Fewer than two results — nothing to compare.
    case single
    /// Top result is clearly separated from the second (topGap >= HIGH_MARGIN).
    case high
    /// Partial separation — some evidence of a best hit.
    case medium
    /// Top results are within epsilon — the list is effectively unranked.
    case low
    /// Query had distinctive tokens (numbers or proper nouns) but no candidate
    /// contained any of them — the recall set is a confident non-match.
    case notFound
}

/// Compute the discrimination level for an ordered score list.
///
/// - Parameter scores: Scores in descending order (highest first). The list
///   must already be sorted; this function does NOT sort.
///
/// Thresholds (heuristic defaults — tune with a real estate; do NOT rank with
/// these values):
///   - `LOW_MARGIN`  = 0.05 — relative gap between rank-1 and rank-2 at which
///     the top result is indistinguishable from rank-2.
///   - `LOW_SPREAD`  = 0.15 — relative gap between rank-1 and the last result
///     at which the entire list is effectively flat.
///   - `HIGH_MARGIN` = 0.25 — relative gap at which rank-1 is clearly the
///     best match.
///   - `EPS`         = 1e-9 — prevents division by zero on all-zero score lists.
public enum RecallDiscrimination {

    // MARK: - Threshold constants (mirrored in recall_discrimination.rs)

    /// Minimum relative gap (rank-1 vs rank-2) for a "not low" classification.
    static let LOW_MARGIN: Double = 0.05
    /// Minimum relative spread (rank-1 vs last) for a "not low" classification.
    static let LOW_SPREAD: Double = 0.15
    /// Minimum relative gap (rank-1 vs rank-2) for a "high" classification.
    static let HIGH_MARGIN: Double = 0.25
    /// Epsilon prevents division by zero on all-zero score vectors.
    static let EPS: Double = 1e-9

    // MARK: - Public API

    /// Classify the discrimination level of an ordered score list.
    public static func classify(_ scores: [Double]) -> DiscriminationLevel {
        guard scores.count >= 2 else { return .single }

        let s0 = scores[0]
        let s1 = scores[1]
        let sLast = scores[scores.count - 1]

        // Denominator: magnitude of the top score, floored at EPS so a
        // zero-top-score list still produces well-defined ratios (all ratios
        // become 0 / EPS ≈ 0 → "low").
        let denom = max(abs(s0), EPS)
        let topGap = (s0 - s1) / denom
        let spread = (s0 - sLast) / denom

        if topGap >= HIGH_MARGIN {
            return .high
        }
        if topGap < LOW_MARGIN && spread < LOW_SPREAD {
            return .low
        }
        return .medium
    }

    // MARK: - Result line

    /// The AI-facing discrimination line to append to a ranked recall result.
    ///
    /// The wording is intentionally factual and action-oriented so the calling
    /// AI knows what to do, not just what the level is.
    ///
    /// When `denseLaneDark` is `true` (the semantic vector lane did not
    /// contribute to this ranking), a `.high` level is capped to `.medium` and
    /// a caveat is appended so the calling AI knows the ranking is lexical-only.
    /// Parity: the same cap and caveat are applied in `recall_discrimination.rs`.
    public static func resultLine(
        for level: DiscriminationLevel,
        denseLaneDark: Bool = false
    ) -> String {
        // Apply the dense-lane-dark cap: "high — clear top result" is
        // misleading when the semantic lane was unavailable. Cap to medium and
        // append a caveat so the calling AI prefers moot_recall_precise.
        let effectiveLevel: DiscriminationLevel
        if denseLaneDark && level == .high {
            effectiveLevel = .medium
        } else {
            effectiveLevel = level
        }

        let base: String
        switch effectiveLevel {
        case .single:
            base = "discrimination: n/a — single/zero results."
        case .high:
            // Only reachable when denseLaneDark == false.
            base = "discrimination: high — clear top result."
        case .medium:
            base = "discrimination: medium — partial separation."
        case .low:
            // For small corpora the semantic/associative modes (conceptual
            // shaped-recall, partial_cue) produce near-flat scores until the
            // embedding encoder (v1.1) lands — this is expected, not an error.
            // Direct the AI toward lexical/precise modes when it needs ranking
            // that it can trust on small estates.
            base = "discrimination: low — top results are within epsilon; treat as effectively unranked. "
                + "Prefer moot_recall_precise / moot_memory_search (ordering: byRelevanceDesc) for "
                + "precision, or widen the query."
        case .notFound:
            // The query carried distinctive tokens (numbers or capitalised words)
            // that are reliable identity markers — yet zero candidates matched
            // any of them. Returning a ranked list here would be fabrication.
            // Direct the AI to try a different query or confirm the content exists.
            base = "discrimination: not_found — query contains distinctive tokens but no stored memory "
                + "matches them. The content may not exist in this estate. "
                + "Try moot_memory_search to confirm, or rephrase the query."
        }

        // Append the dense-lane-dark caveat when the cap fired. The caveat is
        // added regardless of the uncapped level (high → medium cap) so the AI
        // always knows WHY the signal is capped.
        if denseLaneDark && level == .high {
            return base + " (semantic lane dark — ranking is lexical-only; prefer moot_recall_precise.)"
        }
        return base
    }
}
